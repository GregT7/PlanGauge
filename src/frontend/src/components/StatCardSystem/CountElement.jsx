function CountElement({styling, count}) {
    return (
    <div className="flex items-center gap-1">
        <div className={`w-12 h-12 rounded-full ${styling}`}></div>
        <span><span>{`x${count}`}</span></span>
    </div>
)
}

export default CountElement;