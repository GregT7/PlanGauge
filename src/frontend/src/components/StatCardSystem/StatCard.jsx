import {
  Card,
  CardAction,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "@/components/ui/card"
import toLocalMidnight from "@/utils/toLocalMidnight";
import { format } from "date-fns";

function capitalizeFirstLetter(str) {
  if (typeof str !== 'string' || str.length === 0) {
    return str; // Handle non-string or empty inputs
  }
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function determineStatusStyle(status) {
    switch (status) {
        case "neutral":
            return "bg-zinc-900";
        case "good":
            return "bg-emerald-900 border-2 border-emerald-400"
        case "moderate":
            return "bg-amber-900 border-2 border-amber-400"
        case "poor":
            return "bg-rose-900 border-2 border-rose-400";
        default:
            return "bg-red-900 border-2 border-red-500"
    }
}

const StatCard = ({cardData, className=""}) => {
    const cardDate = toLocalMidnight(cardData.date);
    const desciptStr = format(cardDate, 'MMMM d, yyyy')
    
    return (
        <Card className={`${className} ${determineStatusStyle(cardData.status)}`}>
            <CardHeader>
                <CardTitle>{cardData.name}</CardTitle>
                <CardDescription>{desciptStr}</CardDescription>
            </CardHeader>
            <CardContent>
                <p>{`Status: ${capitalizeFirstLetter(cardData.status)}`}</p>
                <p>{`Sum: ${cardData.sum} mins`}</p>
            </CardContent>
        </Card>
    );
}

export default StatCard;