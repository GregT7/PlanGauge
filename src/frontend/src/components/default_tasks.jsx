export default [
  {
    id: 1,
    name: 'Plan Tool p58',
    category: 'Career',
    due_date: Date(2025, 6, 3),
    start_date: Date(2025, 6, 1),
    time_estimation: 90,
    selected: false
  },
  {
    id: 2,
    name: 'Update CV p6',
    category: 'Career',
    due_date: Date(2025, 6, 25),
    start_date: Date(2025, 6, 22),
    time_estimation: 45,
    selected: false
  },
  {
    id: 3,
    name: 'Plan Japan trip p2',
    category: 'Life',
    due_date: Date(2025, 7, 6),
    start_date: Date(2025, 6, 30),
    time_estimation: 90,
    selected: false
  },
];