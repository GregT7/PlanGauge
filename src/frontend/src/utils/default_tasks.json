[
  {
    "id": 1,
    "name": "Plan Tool p58",
    "category": "Career",
    "due_date": "2025-07-03",
    "start_date": "2025-07-01",
    "time_estimation": 90,
    "selected": false
  },
  {
    "id": 2,
    "name": "Update CV p6",
    "category": "Career",
    "due_date": "2025-07-25",
    "start_date": "2025-07-22",
    "time_estimation": 45,
    "selected": false
  },
  {
    "id": 3,
    "name": "Plan Japan trip p2",
    "category": "Life",
    "due_date": "2025-08-06",
    "start_date": "2025-07-30",
    "time_estimation": 90,
    "selected": false
  }
]
