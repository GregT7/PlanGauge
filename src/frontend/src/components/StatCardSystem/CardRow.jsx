import StatCard from "./StatCard"

export default function CardRow ({evaluatedCardStatus, daysRegex, gridClassName, cardData}) {
    return (
        <div className={`${gridClassName} gap-4`}>
            {evaluatedCardStatus
            .filter(card => daysRegex.test(card.name))
            .map(card => (
                <div key={card.date} className="flex justify-center items-center">
                <div className="w-44 h-44 shrink-0">
                    <StatCard cardData={card} />
                </div>
                </div>
            ))}
        </div>
    )
}