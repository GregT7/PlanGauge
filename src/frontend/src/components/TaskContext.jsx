import { createContext, useState } from 'react';
import default_tasks from "./default_tasks";

export const TaskContext = createContext(undefined);


export default function TaskContextProvider({children}) {
    const [tasks, setTasks] = useState(default_tasks);

    return (
        <TaskContext.Provider value={{tasks, setTasks}}>
            {children}
        </TaskContext.Provider>
    );
}