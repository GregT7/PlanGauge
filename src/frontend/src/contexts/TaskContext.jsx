import { createContext, useState } from 'react';
import default_tasks from "../../public/default_tasks";

export const TaskContext = createContext(undefined);


export default function TaskContextProvider({children}) {
    const [tasks, setTasks] = useState(default_tasks);

    return (
        <TaskContext.Provider value={{tasks, setTasks}}>
            {children}
        </TaskContext.Provider>
    );
}