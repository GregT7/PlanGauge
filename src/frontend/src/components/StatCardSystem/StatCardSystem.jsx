import { useContext, useMemo } from 'react';
import { TaskContext } from '@/contexts/TaskContext';
import StatCard from "@/components/StatCardSystem/StatCard";
import StatusCounter from "@/components/StatCardSystem/StatusCounter";
import toLocalMidnight from '@/utils/toLocalMidnight';

const calcStatus = (ave, std, sum) => {
  if (std === 0) return "undetermined";

  const zscore = Math.abs(sum - ave) / std;
  if (zscore <= 1) return "good";
  else if (zscore <= 2) return "moderate";
  else return "poor";
};

function isSameDay(d1, d2) {
  const day1 = toLocalMidnight(d1);
  const day2 = toLocalMidnight(d2);
  return day1.getTime() === day2.getTime();
}

function evalStatus(zScore, _std, sum) {
  zScore = Math.abs(zScore);
  if (sum === 0) return "neutral";
  else if (zScore > 2) return "poor";
  else if (zScore > 1) return "moderate";
  else if (zScore <= 1) return "good";
  else return "undefined";
}

const renderCardRow = (evaluatedCardStatus, daysRegex, gridClassName, cardData) => (
  <div className={`${gridClassName} gap-4`}>
    {evaluatedCardStatus
      .filter(card => daysRegex.test(card.name))
      .map(card => (
        <div key={card.date} className="flex justify-center items-center">
          <div className="w-44 h-44 shrink-0">
            <StatCard cardData={card} />
          </div>
        </div>
      ))}
  </div>
);


function StatCardSystem({cardData}) {
  const { tasks } = useContext(TaskContext);

  const evaluatedCardSums = useMemo(() =>
    cardData.map(card => {
      const cardDate = toLocalMidnight(card.date);
      const daySum = tasks.reduce((sum, task) => {
        const taskDate = toLocalMidnight(task["start_date"]);
        return isSameDay(cardDate, taskDate) ? sum + task["time_estimation"] : sum;
      }, 0);
      return { ...card, sum: daySum };
    }),
    [tasks, cardData]
  );

  const evaluatedCardStatus = useMemo(() =>
    evaluatedCardSums.map(evalCard => {
      const zScore = (evalCard.sum - evalCard.ave) / evalCard.std;
      const newStatus = evalStatus(zScore, evalCard.std, evalCard.sum);
      return { ...evalCard, z_score: zScore, status: newStatus };
    }),
    [evaluatedCardSums]
  );

  const statusCount = useMemo(() => {
    const regex = /^(good|moderate|poor)$/;

    return evaluatedCardStatus.reduce((statObj, task) => {
      const statusValue = task?.status;
      if (statusValue && regex.test(statusValue)) {
        return {
          ...statObj,
          [statusValue]: statObj[statusValue] + 1,
        };
      }
      return statObj;
    }, { good: 0, moderate: 0, poor: 0 });
  }, [evaluatedCardStatus]);

  return (
    <div className="w-[73.5%] mx-auto border-2 border-dashed pt-2">
        <h1 className="text-2xl pt-2 pb-4 text-left pl-8">Stat Card System</h1>

        {renderCardRow(evaluatedCardStatus, /monday|tuesday|wednesday|thursday/i, "grid grid-cols-4", cardData)}
        {renderCardRow(evaluatedCardStatus, /friday|saturday|sunday/i, "grid grid-cols-3 mt-4", cardData)}

        <div>
            <StatusCounter statusCount={statusCount} />
        </div>
    </div>
  );
}

export default StatCardSystem;
