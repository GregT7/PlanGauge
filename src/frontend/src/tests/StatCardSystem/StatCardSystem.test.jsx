import { expect, it, describe } from 'vitest';
import { screen, render } from '@testing-library/react';
import StatCardSystem from '@/components/StatCardSystem/StatCardSystem';
import { useMemo, useState } from 'react';
import { TaskContext } from "@/contexts/TaskContext";
import default_tasks from '@/utils/default_tasks';

const ContextWrapper = ({ initial_tasks, children }) => {
    const [tasks, setTasks] = useState(initial_tasks);

    const timeSum = useMemo(() => 
        tasks.reduce((sum, task) => sum + task.time_estimation, 0),
    [tasks]);

    return (
        <TaskContext.Provider value={{tasks, timeSum}}>
            {children}
        </TaskContext.Provider>
    );
};


// Mock cardData imported inside StatCardSystem
vi.mock('@/utils/testCardData', () => {
    const today = new Date();
    return {
        default: [
            { name: 'Monday', ave: 60, std: 10, sum: 0, date: today },
            { name: 'Tuesday', ave: 30, std: 5, sum: 0, date: today },
            { name: 'Saturday', ave: 50, std: 15, sum: 0, date: today },
    ]
    };
});

const mockTasks = [
    { start_date: new Date(), time_estimation: 60 },
    { start_date: new Date(), time_estimation: 30 },
];

const setup = (tasks = mockTasks) => {
  return render(
    <TaskContext.Provider value={{ tasks }}>
      <StatCardSystem />
    </TaskContext.Provider>
  );
};

describe('StatCardSystem', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });


  it('renders the StatCardSystem component', () => {
    setup();
    // expect(screen.getByText(/Stat Card System/i)).toBeInTheDocument();
    screen.debug()
  });

    it("passes a snapshot test", () => {
        const { container } = render(
            <ContextWrapper initial_tasks={default_tasks}>
                <StatCardSystem />
            </ContextWrapper>
        );

        expect(container).toMatchSnapshot(); // <- actual snapshot test
    });

  it('renders StatCards for days matching regex (weekday and weekend rows)', () => {
    setup();

    // The cards should be rendered based on the mocked cardData
    const cards = screen.getAllByTestId('StatCard');
    expect(cards.length).toBeGreaterThanOrEqual(2); // at least for Mon and Tue
  });

  it('renders the StatusCounter', () => {
    setup();
    expect(screen.getByTestId('StatusCounter')).toBeInTheDocument();
  });

  it('renders correct count bubbles in StatusCounter', () => {
    setup();

    const bubbles = screen.getAllByTestId('circle-div');
    expect(bubbles.length).toBe(3); // good, moderate, poor

    // Optionally check class names for styling
    expect(bubbles[0].className).toMatch(/bg-emerald-500/);
    expect(bubbles[1].className).toMatch(/bg-amber-500/);
    expect(bubbles[2].className).toMatch(/bg-rose-500/);
  });

  it('handles no tasks gracefully (no crash, all statuses should be "neutral")', () => {
    setup([]);

    const cards = screen.getAllByTestId('StatCard');
    expect(cards.length).toBeGreaterThan(0);

    const statusTexts = cards.map(card => card.textContent);
    statusTexts.forEach(text => {
      expect(text.toLowerCase()).toContain('status');
    });
  });

  it('correctly applies z-score classification via integration', () => {
    setup([
      {
        start_date: new Date(),
        time_estimation: 90, // force high z-score for poor classification
      },
    ]);

    const cards = screen.getAllByTestId('StatCard');
    const hasPoor = cards.some(card =>
      card.textContent.toLowerCase().includes('poor')
    );
    expect(hasPoor).toBe(true);
  });
});