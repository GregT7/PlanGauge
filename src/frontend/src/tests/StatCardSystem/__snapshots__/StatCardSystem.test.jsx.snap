// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatCardSystem unit testing > passes a snapshot test 1`] = `
<div>
  <div
    class="w-[73.5%] mx-auto border-2 border-dashed pt-2"
  >
    <h1
      class="text-2xl pt-2 pb-4 text-left pl-8"
    >
      Stat Card System
    </h1>
    <div
      class="grid grid-cols-4 gap-4"
    >
      <div
        class="flex justify-center items-center"
      >
        <div
          class="w-44 h-44 shrink-0"
        >
          <div
            class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm bg-zinc-900 h-full overflow-hidden"
            data-slot="card"
            data-testid="StatCard"
          >
            <div
              class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 space-y-1"
              data-slot="card-header"
            >
              <div
                class="font-semibold text-base truncate"
                data-slot="card-title"
              >
                Monday
              </div>
              <div
                class="text-muted-foreground text-sm truncate"
                data-slot="card-description"
              >
                June 23, 2025
              </div>
            </div>
            <div
              class="px-6 space-y-1 text-sm"
              data-slot="card-content"
            >
              <p
                class="truncate"
              >
                Status: Neutral
              </p>
              <p
                class="truncate"
              >
                Sum: 0 mins
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex justify-center items-center"
      >
        <div
          class="w-44 h-44 shrink-0"
        >
          <div
            class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm bg-zinc-900 h-full overflow-hidden"
            data-slot="card"
            data-testid="StatCard"
          >
            <div
              class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 space-y-1"
              data-slot="card-header"
            >
              <div
                class="font-semibold text-base truncate"
                data-slot="card-title"
              >
                Tuesday
              </div>
              <div
                class="text-muted-foreground text-sm truncate"
                data-slot="card-description"
              >
                June 23, 2025
              </div>
            </div>
            <div
              class="px-6 space-y-1 text-sm"
              data-slot="card-content"
            >
              <p
                class="truncate"
              >
                Status: Neutral
              </p>
              <p
                class="truncate"
              >
                Sum: 0 mins
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid grid-cols-3 mt-4 gap-4"
    >
      <div
        class="flex justify-center items-center"
      >
        <div
          class="w-44 h-44 shrink-0"
        >
          <div
            class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm bg-zinc-900 h-full overflow-hidden"
            data-slot="card"
            data-testid="StatCard"
          >
            <div
              class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 space-y-1"
              data-slot="card-header"
            >
              <div
                class="font-semibold text-base truncate"
                data-slot="card-title"
              >
                Saturday
              </div>
              <div
                class="text-muted-foreground text-sm truncate"
                data-slot="card-description"
              >
                June 23, 2025
              </div>
            </div>
            <div
              class="px-6 space-y-1 text-sm"
              data-slot="card-content"
            >
              <p
                class="truncate"
              >
                Status: Neutral
              </p>
              <p
                class="truncate"
              >
                Sum: 0 mins
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex justify-center gap-5 items-center py-5"
        data-testid="StatusCounter"
      >
        <div
          class="flex items-center gap-1"
        >
          <div
            class="w-12 h-12 rounded-full bg-emerald-500"
            data-testid="circle-div"
          />
          <span>
            <span>
              x0
            </span>
          </span>
        </div>
        <div
          class="flex items-center gap-1"
        >
          <div
            class="w-12 h-12 rounded-full bg-amber-500"
            data-testid="circle-div"
          />
          <span>
            <span>
              x0
            </span>
          </span>
        </div>
        <div
          class="flex items-center gap-1"
        >
          <div
            class="w-12 h-12 rounded-full bg-rose-500"
            data-testid="circle-div"
          />
          <span>
            <span>
              x0
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
